net/tsdial/tsdial.go

```diff
@@ -75,7 +76,8 @@ type Dialer struct {
 	exitDNSDoHBase   string                 // non-empty if DoH-proxying exit node in use; base URL+path (without '?')
 	dnsCache         *dnscache.MessageCache // nil until first non-empty SetExitDNSDoH
 	nextSysConnID    int
-	activeSysConns   map[int]net.Conn // active connections not yet closed
+	activeSysConns   map[int]net.Conn  // active connections not yet closed
+	routes           *bart.Table[bool] // or nil if UserDial should not use routes. `true` indicates routes that point into the Tailscale interface
```

Comment:
grabbing and releasing a mutex here might be too expensive for dns queries. I think you should store this as an `atomic.Pointer`

Reply:
Good point, thank you! Updated to use an `atomic.Pointer`.